### PR TITLE
Feature/client pool

### PR DESCRIPTION
**Pull request by jmaeso**
This PR tries to ~~solve~~ (implement)  #219 . Could be merged but [faktory_worker_go](https://github.com/contribsys/faktory_worker_go/) should be updated by removing `pool.go` and `pool_test.go` and using `github.com/contribsys/faktory/util/pool`.

I would appreciate comments on the PR and feel free to request changes, first time here contributing and didn't feel comfortable with some changes.

I tried not adding breaking changes on this feature.
Didn't add new tests but not sure on how to test this, suggestions are welcome.

### CODE REVIEW

**Comment by mperham** [comment]
Client should be a simple wrapper around socket; it shouldn't know anything about pools.  I would prefer you add Pool to the client package and document how to create a Pool of Clients, which fwg can follow. wdyt?

**Comment by jmaeso** [comment]
So just moving the Pool code to the client side, document how to create a pool of clients, and change the import in the `faktory_worker_go` to point the client package? No real changes to the code?
I thought it was nice to have the thread-safe feature by default (we could also put a limit of connections to 1 as default though), but I guess you have a better idea about the project 😂

**Comment by jbielick** [comment]
I originally wrote the node.js client to be a thin wrapper around a socket, but eventually moved towards running a pool within the client so that the API user could be unaware of the "thread" safety and concurrency benefits while just using the API methods. 

https://github.com/jbielick/faktory_worker_node/blob/d2c6704a267c3c9aea53748c92e1e9e7a45d20db/lib/client.js#L16-L44

I was convinced of this after looking through the Go mysql package and a couple others. It ended up simplifying the code and effectively making the client more reliable / less prone to misuse (thread safety) and not having to educate the user of the client on how to setup and use the connection pool.

> DB is a database handle representing a pool of zero or more underlying connections. It's safe for concurrent use by multiple goroutines.

https://golang.org/pkg/database/sql/#DB

Granted, this object is not called "client", but "DB".

I see benefits to either approach, but I thought I'd weigh in since this was something I grappled with.

**Comment by jmaeso** [comment]
I thought the same as @jbielick, if we move the "pool complexity" in just one place (here in the client), we could also remove the complexity in the [faktory_worker_go](https://github.com/contribsys/faktory_worker_go/blob/master/runner.go#L117) by just having a `faktory.Client` instead of a pool.

If the feature is provided by default to the client then the users don't need to create the "boilerplate" of the pool just to have it concurrent (is hard to see a node.js client being multi-threading by default while the go one isn't :joy: )

I'm just defending the feature "by default" in the client, not defending the code itself, maybe a better implementation would make the functionality look much better (like a renaming `Client` for something like `Service`).

**Comment by mperham** [comment]
You can create a higher-level PooledClient or `client.OpenPool(size)` API which is thread-safe but please keep the current semantics of Client and `client.Open`.  It is typical for APIs to provide both so the caller can choose which type of semantics they need.

**Comment by mperham** [comment]
One hypothetical issue: how do you know that the Pool will return the same Client instance when you write the request and when you read the response?  You need to read/write to the same socket but you aren't holding onto the Client instance during the command transaction so is it possible that you'll write a request to one Client and read the response from a second?  I would refactor the code to minimize Pool checkouts -- hold onto that Client instance during the entire operation.

Can you update test/go_system_test.go to use the new Pooled client so that existing test can verify your changes under heavy load?

**Comment by jmaeso** [comment]
You are completely true, I've missed it. Will be difficult to find all the time for this, but since no one is in a hurry because of this is not a problem.

The first way I've been thinking would involve passing the connection to the write and read functions, but this implies a lot of places [double-casting from closeable interace](https://github.com/contribsys/faktory/pull/222/files#diff-1cd41a2611da49ce63630d5a51bd82c5R401).
You think this casting before any call to read/write functions is good? (doesn't seem to me) 

Would be better to work with [net.Conn interface](https://golang.org/pkg/net/#Conn) in the pool instead of just closeable? (I think should be possible)

**Comment by jmaeso** [comment]
Self-answer: I see the PoolConn definition now... Not sure the best way for the code, will do a quick update for the bug you pointed and lets try to discuss from that point I guess...

File: client/client.go [hunk 1]
```diff
@@ -486,3 +546,18 @@ func hash(pwd, salt string, iterations int) string {
 	}
 	return fmt.Sprintf("%x", hash)
 }
+
+func getTCPConn(pool util.Pool) (*net.TCPConn, error) {
+	poolConn, err := pool.Get()
+	if err != nil {
+		return nil, err
+	}
+	defer poolConn.Close()
```
Comment (mperham) [inline client/client.go]: Aren't you checking in the conn as soon as this function returns, thereby invalidating the connection?  Typically you will use conn within a closure:

getTCPConn(func(conn) {
  conn.Write...
  conn.Read...
})

So the connection is implicitly checked out and checked in for the closure function.

Comment (jmaeso) [reply client/client.go]: That line shouldn't be there at all, my bad (I mean, the pool should not be closed there. In fact, every place that calls this function should `defer tcpConn.Close()`)

But I don't understand what you mean with the closure. Having a valid pool, requesting a connection, doing all operations and then closing that particular connection should be valid, isn't it?

Comment (mperham) [reply client/client.go]: You never close a connection, you check it out, use it to write a request, read a reply and then check it back in.  The pool's Close() call is named poorly; it's actually checking the connection back into the pool, not closing it.

**Comment by mperham** [comment]
Yep, but with a closure the caller doesn’t need to checkout or checkin anything. The closure runner does that for them.

**Pull request closed without merging**